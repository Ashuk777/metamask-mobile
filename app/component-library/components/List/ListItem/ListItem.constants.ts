/* eslint-disable import/prefer-default-export */

// Internal dependencies.
import { VerticalAlignment } from './ListItem.types';

// Defaults
export const DEFAULT_LISTITEM_VERTICALALIGNMENT = VerticalAlignment.Top;
export const DEFAULT_LISTITEM_GAP = 16;
