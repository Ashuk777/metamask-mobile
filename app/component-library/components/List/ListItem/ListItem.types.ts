// Third party dependencies.
import { ViewProps } from 'react-native';

/**
 * Vertical Alignment Options.
 */
export enum VerticalAlignment {
  Top = 'Top',
  Center = 'Center',
  Bottom = 'Bottom',
}

/**
 * ListItem component props.
 */
export interface ListItemProps extends ViewProps {
  /**
   * Content to wrap to display.
   */
  children?: React.ReactNode[] | React.ReactNode;
  /**
   * Optional prop to configure the gap between items inside the ListItem.
   */
  gap?: number | string | undefined;
  /**
   * Optional prop to configure the vertical alignment between items inside the ListItem.
   */
  verticalAlignment?: VerticalAlignment;
}

/**
 * Style sheet input parameters.
 */
export type ListItemStyleSheetVars = Pick<
  ListItemProps,
  'style' | 'verticalAlignment'
>;
