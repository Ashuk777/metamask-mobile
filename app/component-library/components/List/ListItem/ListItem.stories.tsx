/* eslint-disable no-console */

// Third party dependencies.
import React from 'react';
import { storiesOf } from '@storybook/react-native';

// External dependencies.
import Button, { ButtonVariants } from '../../Buttons/Button';

// Internal dependencies.
import ListItem from './ListItem';
import { ListItemProps } from './ListItem.types';

const ListItemStory = () => (
  <ListItem>
    <Button
      variant={ButtonVariants.Secondary}
      label="Test"
      onPress={() => console.log('clicked')}
    />
    <Button
      variant={ButtonVariants.Secondary}
      label="Hello"
      onPress={() => console.log('clicked')}
    />
  </ListItem>
);

storiesOf('Morph / Lists', module).add('ListItem', ListItemStory);

export default ListItemStory;
