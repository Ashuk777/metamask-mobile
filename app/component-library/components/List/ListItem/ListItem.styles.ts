// Third party dependencies.
import { StyleSheet, ViewStyle } from 'react-native';

// External dependencies.
import { Theme } from '../../../../util/theme/models';

// Internal dependencies.
import { VerticalAlignment, ListItemStyleSheetVars } from './ListItem.types';

/**
 * Style sheet function for ListItem component.
 *
 * @param params Style sheet params.
 * @param params.theme App theme from ThemeContext.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: { theme: Theme; vars: ListItemStyleSheetVars }) => {
  const { vars } = params;
  const { style, verticalAlignment } = vars;
  let alignItems;
  switch (verticalAlignment) {
    case VerticalAlignment.Top:
      alignItems = 'flex-start';
      break;
    case VerticalAlignment.Center:
      alignItems = 'center';
      break;
    case VerticalAlignment.Bottom:
      alignItems = 'flex-end';
      break;
    default:
      alignItems = 'flex-start';
  }

  return StyleSheet.create({
    base: Object.assign(
      {
        flexDirection: 'row',
        alignItems,
      } as ViewStyle,
      style,
    ) as ViewStyle,
  });
};

export default styleSheet;
