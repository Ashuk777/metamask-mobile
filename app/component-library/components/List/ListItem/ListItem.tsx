/* eslint-disable react/prop-types */

// Third party dependencies.
import React from 'react';
import { View } from 'react-native';

// External dependencies.
import { useStyles } from '../../../hooks';

// Internal dependencies.
import styleSheet from './ListItem.styles';
import { ListItemProps } from './ListItem.types';
import {
  DEFAULT_LISTITEM_PADDING,
  DEFAULT_LISTITEM_BORDERRADIUS,
  DEFAULT_LISTITEM_GAP,
  DEFAULT_LISTITEM_VERTICALALIGNMENT,
  TESTID_LISTITEM,
  TESTID_LISTITEM_GAP,
} from './ListItem.constants';

const ListItem: React.FC<ListItemProps> = ({
  style,
  children,
  padding = DEFAULT_LISTITEM_PADDING,
  borderRadius = DEFAULT_LISTITEM_BORDERRADIUS,
  gap = DEFAULT_LISTITEM_GAP,
  verticalAlignment = DEFAULT_LISTITEM_VERTICALALIGNMENT,
}) => {
  const { styles } = useStyles(styleSheet, {
    style,
    padding,
    borderRadius,
    verticalAlignment,
  });
  return (
    <View style={styles.base} testID={TESTID_LISTITEM}>
      {React.Children.map(children, (child, index) => (
        <>
          {index > 0 && (
            <View style={{ width: gap }} testID={TESTID_LISTITEM_GAP} />
          )}
          {child}
        </>
      ))}
    </View>
  );
};

export default ListItem;
