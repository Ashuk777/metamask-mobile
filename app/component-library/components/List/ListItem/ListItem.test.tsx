// Third party dependencies.
import React from 'react';
import { shallow } from 'enzyme';
import { View } from 'react-native';

// External dependencies.

// Internal dependencies.
import ListItem from './ListItem';
import {
  DEFAULT_LISTITEM_PADDING,
  DEFAULT_LISTITEM_BORDERRADIUS,
  DEFAULT_LISTITEM_GAP,
  TESTID_LISTITEM,
  TESTID_LISTITEM_GAP,
} from './ListItem.constants';
import { VerticalAlignment } from './ListItem.types';

describe('ListItem', () => {
  it('should render snapshot correctly', () => {
    const wrapper = shallow(
      <ListItem>
        <View />
      </ListItem>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render component correctly', () => {
    const wrapper = shallow(
      <ListItem>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.exists()).toBe(true);
  });

  it('should render the correct default padding', () => {
    const wrapper = shallow(
      <ListItem>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.props().style.padding).toBe(
      DEFAULT_LISTITEM_PADDING,
    );
  });

  it('should render the given padding', () => {
    const givenPadding = 12;
    const wrapper = shallow(
      <ListItem padding={givenPadding}>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.props().style.padding).toBe(givenPadding);
  });

  it('should render the correct default borderRadius', () => {
    const wrapper = shallow(
      <ListItem>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.props().style.borderRadius).toBe(
      DEFAULT_LISTITEM_BORDERRADIUS,
    );
  });

  it('should render the given borderRadius', () => {
    const givenBorderRadius = 12;
    const wrapper = shallow(
      <ListItem borderRadius={givenBorderRadius}>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.props().style.borderRadius).toBe(
      givenBorderRadius,
    );
  });

  it('should render the correct default gap', () => {
    const wrapper = shallow(
      <ListItem>
        <View />
        <View />
      </ListItem>,
    );
    const listItemGapComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM_GAP,
    );
    expect(listItemGapComponent.props().style.width).toBe(DEFAULT_LISTITEM_GAP);
  });

  it('should render the given gap', () => {
    const givenGap = 14;
    const wrapper = shallow(
      <ListItem gap={givenGap}>
        <View />
        <View />
      </ListItem>,
    );
    const listItemGapComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM_GAP,
    );
    expect(listItemGapComponent.props().style.width).toBe(givenGap);
  });

  it('should render the correct default verticalAlignment', () => {
    const wrapper = shallow(
      <ListItem>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.props().style.alignItems).toBe('flex-start');
  });

  it('should render the given verticalAlignment', () => {
    const wrapper = shallow(
      <ListItem verticalAlignment={VerticalAlignment.Center}>
        <View />
      </ListItem>,
    );
    const listItemComponent = wrapper.findWhere(
      (node) => node.prop('testID') === TESTID_LISTITEM,
    );
    expect(listItemComponent.props().style.alignItems).toBe('center');
  });
});
