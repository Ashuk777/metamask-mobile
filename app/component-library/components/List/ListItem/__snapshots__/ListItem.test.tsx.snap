// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render snapshot correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "borderRadius": 0,
      "flexDirection": "row",
      "padding": 16,
    }
  }
  testID="listitem"
>
  <View />
</View>
`;
